### PR TITLE
Add methods for ABI encoding/decoding circuit inputs

### LINKED ISSUE #453: ABI encoding functions should be exposed by `noirc_abi` crate.
# Problem

Given a compiled circuit, I'd like to be able to encode the inputs to prove/verify proofs against it. Nargo contains functions to do this however I have to use the CLI to access them whereas I'd like to do this from within code.

# Solution

It seems natural that the `noirc_abi` crate should expose a way to abi encode arguments in order to feed it into a proof/verification function. We should move the relevant functions to there and the nargo CLI (and any other crates) can then just import them from there.

# Alternatives considered

I could copy paste this code into my project but this adds overhead of maintenance going forwards.

# Additional context

(If applicable.)

### LINKED ISSUE #545: Standardise ABI encoding implementation for proving/verifying.
# Problem

When converting inputs into values which are to be passed to the backend, there's two things we need to do:
1. Convert the `InputValue`s to `FieldElement`s (ABI encoding)
2. Write the `FieldElement`s into the format expected by the backend.

We currently do both of these in a single step for proving and verifying:
- `process_abi_with_input`: `BTreeMap<String, InputValue>` -> `BTreeMap<Witness, FieldElement>`
- `process_abi_with_verifier_input`: `BTreeMap<String, InputValue>` -> `Vec<FieldElement>` 

This means we need to update both of these functions together whenever a new type is added, etc., increasing the chances of mistakes.

# Solution

Step 1 doesn't depend on whether we're proving or verifying so we should be able to have a single implementation for both.

The steps for proving/verifying are then:
- Proving: `BTreeMap<String, InputValue>` -> `Vec<FieldElement>` -> `BTreeMap<Witness, FieldElement>`
- Verifying: `BTreeMap<String, InputValue>` -> `Vec<FieldElement>` 

The only difference between the two cases is then that we need to do the relatively simple step of writing the abi encoded inputs into the witness map when proving.

# Alternatives considered

This depends on the assumption that we can write the `Vec<FieldElement>` into the witness map without any knowledge of the ABI (currently we just copy the elements of the vector to witnesses 1..N+1). If this doesn't hold in future then we'd need to create the witness map while processing the ABI.

# Additional context

(If applicable.)

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

closes #545 
closes #453 

# Description

## Summary of changes

This PR defines `abi.encode` and `abi.decode` methods which convert between `BTreeMap<String, InputValue>`  and `Vec<FieldElement>`

`abi.encode` allows us to replace `process_abi_with_input` and `process_abi_with_verifier_input` with a shared implementation. This makes it easier to support new types in future and allows for better testing.

`abi.decode` allows us to simplify the code for exporting public inputs to `Verifier.toml`. Instead of having to pass around the witness index which corresponds to the return value and pull it out of the witness map then merge with the rest of the inputs, we can in a single step decode the entirety of the circuits inputs from the witness map.

## Dependency additions / changes

(If applicable.)

## Test additions / changes

(If applicable.)

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

# Additional context

Moving the handling of inputs to the `noirc_abi` crate also makes it easier to reuse this implementation through wasm in future.

### CODE REVIEW

File: crates/nargo/src/cli/prove_cmd.rs [hunk 1]
```diff
@@ -167,27 +61,51 @@ pub fn parse_and_solve_witness<P: AsRef<Path>>(
...
+    })?;
+
+    let mut solved_witness: BTreeMap<Witness, FieldElement> = encoded_inputs
+        .into_iter()
+        .enumerate()
+        .map(|(index, witness_value)| {
+            let witness = Witness::new(WITNESS_OFFSET + (index as u32));
+            (witness, witness_value)
+        })
+        .collect();
```
Comment (TomAFrench) [inline crates/nargo/src/cli/prove_cmd.rs:108]: Ideally we'd have something like `compiled_program.circuit.inputs` so we could write in the inputs without needing to know about the witness offset similarly to how we're extracting the public inputs.

File: crates/noirc_abi/src/lib.rs [hunk 2]
```diff
@@ -100,13 +105,157 @@ impl Abi {
     pub fn num_parameters(&self) -> usize {
         self.parameters.len()
     }
+
+    /// Returns the number of field elements required to represent the ABI's input once encoded.
+    pub fn field_count(&self) -> u32 {
+        self.parameters.iter().fold(0, |acc, (_, param_type)| acc + param_type.field_count())
```
Comment (jfecher) [inline crates/noirc_abi/src/lib.rs:111]: Can also use `self.parameters.iter().map(|(_, typ)| typ.field_count()).sum()`

**Review by jfecher — APPROVED** [review]

File: crates/noirc_abi/src/lib.rs (hunk 2 above)
Comment (TomAFrench) [reply crates/noirc_abi/src/lib.rs:111]:
```suggestion
        self.parameters.iter().map(|(_, typ)| typ.field_count()).sum()
```

**Review by jfecher — APPROVED** [review]

File: crates/noirc_abi/src/lib.rs [hunk 3]
```diff
@@ -100,13 +105,157 @@ impl Abi {
...
+            // As the circuit calculates the return value in the process of calculating rest of the witnesses
+            // it's not absolutely necessary to provide them as inputs. We then tolerate an undefined value for
+            // the return value input and just skip it.
+            if allow_undefined_return
+                && param_name == MAIN_RETURN_NAME
+                && matches!(value, InputValue::Undefined)
+            {
+                let return_witness_len = param_type.field_count();
+
+                // We do not support undefined arrays for now - TODO
```
Comment (kevaundray) [inline crates/noirc_abi/src/lib.rs:154]: Can you open up an issue for this TODO once merged?

Comment (jfecher) [reply crates/noirc_abi/src/lib.rs:154]: This is an existing TODO from when the inferred return value from main feature was first added IIRC

Comment (kevaundray) [reply crates/noirc_abi/src/lib.rs:154]: Ahh okay, I'll open the issue for this once merged, as it was not Tom's addition

**Pull request merged**